chain/chain/src/chain.rs

```diff
@@ -200,6 +216,37 @@ impl OrphanBlockPool {
 
         ret
     }
+
+    pub fn get_orphans_within_depth(
+        &self,
+        prev_hash: CryptoHash,
+        target_depth: u64,
+    ) -> Vec<CryptoHash> {
+        let mut visited = HashSet::new();
+        let mut queue = vec![(prev_hash, 0)];
+        while let Some((prev_hash, depth)) = queue.pop() {
+            if depth == target_depth {
+                continue;
+            }
+            if let Some(block_hashes) = self.prev_hash_idx.get(&prev_hash) {
+                for hash in block_hashes {
+                    // there should be no loop, but just in case
+                    if visited.insert(*hash) {
+                        queue.push((*hash, depth + 1));
```

Comment:
Actually since this property must hold (otherwise something is seriously wrong), I suggest that we avoid this checking and add some `debug_assert` to further improve performance. This should also allow us to change visited to `Vec` (which also slightly improves performance)